src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+                    return;
+                }
+
+                try
+                {
+                    this.Task.Run(this);
+                }
+                catch (Exception t)
+                {
+                    if (Logger.WarnEnabled)
+                    {
+                        Logger.Warn($"An exception was thrown by {this.Task.GetType().Name}.", t);
+                    }
+                }
+            }
+
+            public override string ToString()
+            {
+                PreciseTimeSpan currentTime = PreciseTimeSpan.FromStart - this.timer.StartTime;
+                TimeSpan remaining = this.Deadline - currentTime.ToTimeSpan();
+
+                StringBuilder buf = new StringBuilder(192)
+                    .Append(this.GetType().Name)
+                    .Append('(')
+                    .Append("deadline: ");
+                if (remaining > TimeSpan.Zero)
+                {
+                    buf.Append(remaining)
+                        .Append(" later");
+                }
+                else if (remaining < TimeSpan.Zero)
+                {
+                    buf.Append(-remaining)
+                        .Append(" ago");
+                }
+                else
+                {
+                    buf.Append("now");
+                }
+
+                if (this.Cancelled)
+                {
+                    buf.Append(", cancelled");
+                }
+
+                return buf.Append(", task: ")
+                    .Append(this.Task)
+                    .Append(')')
+                    .ToString();
+            }
+        }
+
+        /// <summary>
+        /// Bucket that stores HashedWheelTimeouts. These are stored in a linked-list like datastructure to allow easy
+        /// removal of HashedWheelTimeouts in the middle. Also the HashedWheelTimeout act as nodes themself and so no
+        /// extra object creation is needed.
+        /// </summary>
+        sealed class HashedWheelBucket
+        {
+            // Used for the linked-list datastructure
+            HashedWheelTimeout head;
+            HashedWheelTimeout tail;
+
+            /// <summary>
+            /// Add {@link HashedWheelTimeout} to this bucket.
+            /// </summary>
+            public void AddTimeout(HashedWheelTimeout timeout)
+            {
+                Contract.Assert(timeout.Bucket == null);
+                timeout.Bucket = this;
+                if (this.head == null)
+                {
+                    this.head = this.tail = timeout;
+                }
+                else
+                {
+                    this.tail.Next = timeout;
+                    timeout.Prev = this.tail;
+                    this.tail = timeout;
+                }
+            }
+
+            /// <summary>
+            /// Expire all <see cref="HashedWheelTimeout"/>s for the given <c>deadline</c>.
+            /// </summary>
+            public void ExpireTimeouts(TimeSpan deadline)
```

Comment:
we expect this to be efficient because we expect most timers to be canceled before it comes back to the bucket again. would be good to get a trace if the assumption we made is incorrect (in which case, user would tune this thing to have more buckets or a greater tick duration potentially)